integration-tests/test/mrf_lp.spec.ts

```diff
@@ -1707,9 +1706,606 @@ describe('Liquidity Pool Test', async () => {
+          depositAmount.mul(remainingERC20Percent).div(1000)
+        )
+      )
+    })
+
+    it('should revert unfulfillable swap-ons in batch', async () => {
+      const depositAmount = utils.parseEther('10')
+
+      await approveTransaction(
+        L1ERC20_1,
+        L1LiquidityPool.address,
+        depositAmount
+      )
+
+      const preL2EthBalance = await env.l2Wallet.getBalance()
+      const preL2ERC20_1Balance = await L2ERC20_1.balanceOf(
+        env.l2Wallet.address
+      )
+      const preL2ERC20_2Balance = await L2ERC20_2.balanceOf(
+        env.l2Wallet.address
+      )
+
+      const userRewardFeeRate = await L1LiquidityPool.getUserRewardFeeRate(
+        ethers.constants.AddressZero
+      )
+      const remainingETHPercent = await getRemainingPercent(userRewardFeeRate)
+      const userRewardERC20_1FeeRate =
+        await L2LiquidityPool.getUserRewardFeeRate(L2ERC20_1.address)
+      const userRewardERC20_2FeeRate = await L1LiquidityPool.getUserRewardFeeRate(L1ERC20_2.address)
+      const remainingERC20_2Percent = await getRemainingPercent(
+        userRewardERC20_2FeeRate
+      )
+
+      const requestedETHLiquidity = (
+        await env.l2Provider.getBalance(L2LiquidityPool.address)
+      ).add(ethers.utils.parseEther('10'))
+      const swapOnETHAmount = requestedETHLiquidity
+        .mul(1000)
+        .div(remainingETHPercent)
+
+      const requestedERC20Liquidity = (
+        await L2ERC20_2.balanceOf(L2LiquidityPool.address)
+      ).add(ethers.utils.parseEther('10'))
+      const swapOnERC20Amount = requestedERC20Liquidity
+        .mul(1000)
+        .div(remainingERC20_2Percent)
+
+      await approveTransaction(
+        L1ERC20_2,
+        L1LiquidityPool.address,
+        swapOnERC20Amount
+      )
+
+      await env.waitForRevertXDomainTransaction(
+        L1LiquidityPool.clientDepositL1Batch(
+          [
+            {
+              amount: swapOnETHAmount,
+              l1TokenAddress: ethers.constants.AddressZero,
+            },
+            { amount: depositAmount, l1TokenAddress: L1ERC20_1.address },
+            { amount: swapOnERC20Amount, l1TokenAddress: L1ERC20_2.address },
+          ],
+          { value: swapOnETHAmount, gasLimit: 9000000 }
+        ),
+        Direction.L1ToL2
+      )
+
+      const remainingERC20_1Percent = await getRemainingPercent(
+        userRewardERC20_1FeeRate
+      )
+
+      const postL2EthBalance = await env.l2Wallet.getBalance()
+      const postL2ERC20_1Balance = await L2ERC20_1.balanceOf(
+        env.l2Wallet.address
+      )
+      const postL2ERC20_2Balance = await L2ERC20_2.balanceOf(
+        env.l2Wallet.address
+      )
+
+      expect(preL2EthBalance).to.deep.eq(postL2EthBalance)
+      expect(postL2ERC20_1Balance).to.deep.eq(
+        preL2ERC20_1Balance.add(
+          depositAmount.mul(remainingERC20_1Percent).div(1000)
+        )
       )
+      expect(preL2ERC20_2Balance).to.deep.eq(postL2ERC20_2Balance)
     })
   })
 })
```

Comment:
minor: could add another test to verify a batch consisting of all same tokens

Reply:
```
    it('should deposit ERC20 twice in batch', async () => {
      const depositAmount = utils.parseEther('10')

      const preL1ERC20Balance = await L1ERC20_1.balanceOf(env.l1Wallet.address)
      const preL2ERC20Balance = await L2ERC20_1.balanceOf(env.l2Wallet.address)

      await approveTransaction(
        L1ERC20_1,
        L1LiquidityPool.address,
        depositAmount.mul(BigNumber.from(2))
      )
      const userRewardERC20FeeRate = await L2LiquidityPool.getUserRewardFeeRate(
        L2ERC20_1.address
      )

      const depositTx = await env.waitForXDomainTransaction(
        L1LiquidityPool.clientDepositL1Batch([
          { amount: depositAmount, l1TokenAddress: L1ERC20_1.address },
          { amount: depositAmount, l1TokenAddress: L1ERC20_1.address },
        ]),
        Direction.L1ToL2
      )

      const remainingERC20Percent = await getRemainingPercent(
        userRewardERC20FeeRate
      )

      const postL1ERC20Balance = await L1ERC20_1.balanceOf(env.l1Wallet.address)
      const postL2ERC20Balance = await L2ERC20_1.balanceOf(env.l2Wallet.address)

      expect(preL1ERC20Balance).to.deep.eq(
        postL1ERC20Balance.add(depositAmount.mul(2))
      )
      expect(postL2ERC20Balance).to.deep.eq(
        preL2ERC20Balance.add(
          depositAmount.mul(2).mul(remainingERC20Percent).div(1000)
        )
      )
    })

    it('should deposit ETH twice in batch', async () => {
      const depositAmount = utils.parseEther('10')

      const preL2ETHBalance = await env.l2Wallet.getBalance()
      const userRewardETHFeeRate = await L2LiquidityPool.getUserRewardFeeRate(
        env.ovmEth.address
      )

      const depositTx = await env.waitForXDomainTransaction(
        L1LiquidityPool.clientDepositL1Batch(
          [
            {
              amount: depositAmount,
              l1TokenAddress: ethers.constants.AddressZero,
            },
            {
              amount: depositAmount,
              l1TokenAddress: ethers.constants.AddressZero,
            },
          ],
          { value: depositAmount.mul(2) }
        ),
        Direction.L1ToL2
      )

      const remainingETHPercent = await getRemainingPercent(
        userRewardETHFeeRate
      )
      const postL2ETHBalance = await env.l2Wallet.getBalance()

      expect(postL2ETHBalance).to.deep.eq(
        preL2ETHBalance.add(
          depositAmount.mul(2).mul(remainingETHPercent).div(1000)
        )
      )
    })
```